libr/cons/canvas.c

```diff
@@ -261,47 +160,214 @@ static const char *set_attr(RConsCanvas *c, const char *s) {
 	return p;
 }
 
-R_API void r_cons_canvas_write(RConsCanvas *c, const char *s) {
-	char *p, ch;
-	int orig_x, x;
-	int left, slen, attr_len, piece_len;
+R_API bool r_cons_canvas_gotoxy(RConsCanvas *c, int x, int y) {
+	bool ret = true;
+	if (!c) {
+		return 0;
+	}
+	y += c->sy;
+	x += c->sx;
+
+	if (x > c->blen[y] * 2) {
+		return false;
+	}
+	if (y > c->h * 2) {
+		return false;
+	}
+	if (x >= c->blen[y]) {
+		c->x = c->blen[y];
+		ret = false;
+	}
+	if (y >= c->h) {
+		c->y = c->h;
+		ret = false;
+	}
+	if (x < 0) {
+		//c->x = 0;
+		ret = false;
+	}
+	if (y < 0) {
+		c->y = 0;
+		ret = false;
+	}
+	if (x < c->blen[y] && x >= 0) {
+		c->x = x;
+	}
+	if (y < c->h && y >= 0) {
+		c->y = y;
+	}
+	return ret;
+}
+
+R_API RConsCanvas *r_cons_canvas_new(int w, int h) {
+	if (w < 1 || h < 1) {
+		return NULL;
+	}
+	RConsCanvas *c = R_NEW0 (RConsCanvas);
+	if (!c) {
+		return NULL;
+	}
+	c->color = 0;
+	c->sx = 0;
+	c->sy = 0;
+	c->b = malloc (sizeof *c->b * h);
+	if (!c->b) {
+		goto error_b;
+	}
+	c->blen = malloc (sizeof *c->blen * h);
+	if (!c->blen) {
+		goto error_blen;
+	}
+	c->bsize = malloc (sizeof *c->bsize * h);
+	if (!c->bsize) {
+		goto error_bsize;
+	}
+	int i;
+	for (i = 0; i < h; i++) {
+		c->b[i] = malloc (w + 1);
+		c->blen[i] = w;
+		c->bsize[i] = w + 1;
+		if (!c->b[i]) {
+			goto error_c;
+		}
+	}
+	c->w = w;
+	c->h = h;
+	c->x = c->y = 0;
+	c->attrslen = 0;
+	c->attrs = calloc (sizeof (*c->attrs), (c->w + 1) * c->h);
+	if (!c->attrs) {
+		goto error_c;
+	}
+	c->attr = Color_RESET;
+	r_cons_canvas_clear (c);
+	return c;
+error_c: {
```

Comment:
as long as its initialized to 0, you can safely call free on all the fields without having to add many labels and gotos. just 1

Reply:
in gstreamer.. and eventually in r2land, we name those labels as locations. Like `beach`. so you can `goto beach;`